### PR TITLE
feat(spanner): set resource header in data client

### PR DESCRIPTION
**Pull request by skuruppu**
Sets the google-cloud-resource-prefix header in the Spanner data client so that we can support resource-based routing.

Towards #5181

### CODE REVIEW

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    public partial class SpannerClientImpl
+    {
+        /// <summary>
+        /// The name of the header used for efficiently routing requests.
+        /// </summary>
+        /// <remarks>
+        /// This should be set to the database resource name ("projects/{projectId}/instances/{instanceId}/databases/{databaseId}") for any RPC.
+        /// For non-streaming calls, <see cref="SpannerClientImpl"/> performs this automatically. This cannot be performed
+        /// automatically for streaming calls due to the separation between initializing the stream and sending requests, so
+        /// client code should set the value in a <see cref="CallSettings"/>. Typically this is performed with either the
```
Comment (skuruppu) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs:31]: This part worries me, I think I have to do this in the hand-written later.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+        // Visible for testing
+
+        /// <summary>
+        /// Retrieves the database resource name from a full resource name.
+        /// Validation is performed as far as the database ID but no further; the database ID is deemed to end at the first slash.
+        /// </summary>
+        /// <param name="resource">The resource name, which must start with projects/{project_id}/instances/{instance_id}/databases/{database_id}</param>
+        /// <returns>The database resource name</returns>
+        internal static string GetDatabaseResourceName(string resource)
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs:95]: Drive-by as I don't have time to review *right* now - I strongly suspect this could be massively reduced using DatabaseName.TryParse or DatabaseName.Parse.

Comment (skuruppu) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs:95]: Ah yes, that's a good idea. I will fix it early next week. Don't worry about reviewing until I fix it.

Comment (skuruppu) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs]: That was much nicer. I still want to write more comprehensive tests for whether the header is set but at least logic-wise, this is hopefully now closer to the final implementation.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/SpannerClientImplTest.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        [Fact]
+        public void ClientSetsResourcePrefixHeader()
+        {
+            var invoker = new FakeCallInvoker();
+            var client = new SpannerClientBuilder { CallInvoker = invoker }.Build();
+            const string session = "projects/proj/instances/inst/databases/db/sessions/sess";
+            client.BeginTransaction(new BeginTransactionRequest { Session = session });
+            var found = false;
+            foreach (Metadata.Entry e in invoker.Metadata)
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/SpannerClientImplTest.cs:31]: I'd suggest using:

```csharp
Metadata.Entry entry = invoker.Metadata.Single(e => e.Key == "google-cloud-resource-prefix");
Assert.Equal("projects/proj/instances/inst/databases/db", entry.Value);
```

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs:32]: Yes, that would make sense. There are a few places we could do it - I don't think it'll be terribly hard. I'm happy for you to look, or I can do so. I'd suggest doing that as a separate PR, just for simplicity.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            if (string.IsNullOrEmpty(resource))
+            {
+                return;
+            }
+
+            // Try to parse the resource as a database name.
+            DatabaseName database;
+            if (!DatabaseName.TryParse(resource, out database))
+            {
+                // If that fails, the resource name is likely a session name.
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs:89]: Do we definitely want to use Parse rather than TryParse here? If we have a call with a resource which is neither a session or a database name, do we definitely want it to throw here?

Moreover, might it not make sense to use different methods here, one expecting a session name and one expecting a database name? I'd expect us to know which should be used on a per-request basis.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/SpannerClientImplTest.cs (hunk 3 above)
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/SpannerClientImplTest.cs:31]: Hmmm I was trying to do something similar by following this [example](https://github.com/googleapis/google-cloud-dotnet/blob/d1056b9a083ab925bbe63053624b79f108d0c3ee/apis/Google.Cloud.Language.V1/Google.Cloud.Language.V1.Tests/ApiClientHeaderTest.cs#L32). Except the compiler kept telling me that `Metadata` doesn't contain a definition for `FirstOrDefault` and `Single`.

I looked at the [Metadata docs](https://grpc.github.io/grpc/csharp-dotnet/api/Grpc.Core.Metadata.html#methods) and indeed those methods are not defined. Then I thought I would use `Metadata.Contains` but xUnit was unhappy about that.

Some Googling gave me the idea to use Assert.Contains() which finally worked :)

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs (hunk 1 above)
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs:32]: Yup, happy to put this in a separate PR.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs (hunk 4 above)
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientPartial.cs:89]: I guess you're right and it's overkill to fail the request here if an appropriate resource name is not specified. I guess that's not the job of the client library.

I split the function into two as you suggested and now ignoring the failure to parse.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/SpannerClientImplTest.cs (hunk 3 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/SpannerClientImplTest.cs:31]: `Metadata` implements `IList<Metadata.Entry>`. `Single` is an extension method defined in LINQ, so `using System.Linq;` would have worked.

However, thinking about it, `Assert.Single` is *really* what you want:

```csharp
Metadata.Entry entry = Assert.Single(invoker.Metadata, e => e.Key == "google-cloud-resource-prefix");
Assert.Equal("projects/proj/instances/inst/databases/db", entry.Value);
```

That way you'll get a better error message if there are no/multiple matches.

Comment (skuruppu) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/SpannerClientImplTest.cs]: You're right. We don't want to set it multiple times. Updated the implementation and also added tests for all methods. There one (ListSessions) that I can't get working so I will look into it tomorrow.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/ResourcePrefixHeaderTest.cs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+            Assert.Equal(database, entry.Value);
+        }
+
+        [Fact]
+        public void SetsHeaderOnListSessions()
+        {
+            var invoker = new FakeCallInvoker();
+            var client = new SpannerClientBuilder { CallInvoker = invoker }.Build();
+            client.ListSessions(new ListSessionsRequest { Database = database });
+            // TODO(skuruppu): Figure out why the invoker doesn't have Metadata set.
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/ResourcePrefixHeaderTest.cs:64]: I suspect this is because it hasn't actually issued a request yet. List operations are lazy. If you change the above to 

```csharp
client.ListSessions(new ListSessionsRequest { Database = database }).AsRawResponses().First();
```

... then I suspect it'll work. (Or it may throw another exception because your fake doesn't populate the response appropriately... you'll have to see.)

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/ResourcePrefixHeaderTest.cs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+// limitations under the License.
+
+using Grpc.Core;
+using System;
+using System.Linq;
+using Xunit;
+
+namespace Google.Cloud.Spanner.V1.Tests
+{
+    public class ResourcePrefixHeaderTest
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/ResourcePrefixHeaderTest.cs:22]: I think you should add a couple of tests (one for database, one for session) for cases where the request contains a badly-formatted resource name. (And another for when it's not populated at all.)

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/ResourcePrefixHeaderTest.cs [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    {
+        const string header = "google-cloud-resource-prefix";
+        const string database = "projects/proj/instances/inst/databases/db";
+        const string session = "projects/proj/instances/inst/databases/db/sessions/sess";
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/ResourcePrefixHeaderTest.cs:26]:
```suggestion
        private const string ResourcePrefixHeader = "google-cloud-resource-prefix";
        private const string SampleDatabaseName = "projects/proj/instances/inst/databases/db";
        private const string SampleSessionName = "projects/proj/instances/inst/databases/db/sessions/sess";
```

(If you're using Visual Studio, use F2 to do the renaming and all the uses will change too.)

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/ResourcePrefixHeaderTest.cs (hunk 5 above)
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/ResourcePrefixHeaderTest.cs:64]: Thanks heaps for the suggestion, it works now!

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/ResourcePrefixHeaderTest.cs (hunk 6 above)
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1.Tests/ResourcePrefixHeaderTest.cs:22]: That's a good idea. Sorry I missed this. I added 4 tests.

**Comment by skuruppu** [comment]
PTAL @jskeet. I've addressed all the feedback and got that last test to pass.

**Comment by skuruppu** [comment]
Friendly ping on this PR @jskeet. Just curious to know if there's anything else I should do here. Otherwise I'll move onto supporting the header in streaming RPCs.

**Comment by skuruppu** [comment]
All good, no worries at all @jskeet. Thanks for merging it in. I'll start work on the next piece.
